### PR TITLE
chore: update docker to v25

### PR DESCRIPTION
**Pull request by matejvasek**
update docker to v25

### CODE REVIEW

**Comment by matejvasek** [comment]
@coreydaley

File: pkg/docker/docker.go
Comment (matejvasek) [inline pkg/docker/docker.go:324]: Are these even need today? AFAIK just using `dockerapi.FromEnv` should ensure everything is set up just all right.

**Comment by matejvasek** [comment]
@coreydaley is the security check failure really related to my PR?

**Comment by coreydaley** [comment]
@matejvasek No, and the security check is also option, so it will not stop the pull request from merging.

**Comment by matejvasek** [comment]
@coreydaley We need this to get rid off some CVE reports in RH jira. The CVEs do not really affects us but we do not want to see the warnings.

File: pkg/docker/docker.go
Comment (adambkaplan) [inline pkg/docker/docker.go:325]: My understanding is that `APIVersionNegotiation` is more or less incompatible with `FromEnv` if the `DOCKER_API_VERSION` env var is set.

See https://github.com/moby/moby/blob/24.0/client/client_deprecated.go#L13-L17

**Review by adambkaplan — CHANGES_REQUESTED** [review]
The update to the new client constructor is not a clean port from v24 to v25. I'm concerned that this is exercised in the `s2i build` code paths, which we don't have great test coverage for at present. OpenShift and the s2i container images for Tekton, Shipwright, and KNative use the `s2i generate` code paths, which we can verify in CI.

File: pkg/docker/docker.go
Comment (matejvasek) [reply pkg/docker/docker.go:325]: I think `WithAPIVersionNegotiation()` is necessary. It does not work on my system with older moby.
Of course if `DOCKER_API_VERSION` is set that will take precedence.

File: pkg/docker/docker.go
Comment (matejvasek) [reply pkg/docker/docker.go:326]: FromEnv probably takes more envvars into account witch is better, or not?

File: pkg/docker/docker.go
Comment (matejvasek) [reply pkg/docker/docker.go:325]: so `DOCKER_API_VERSION` beats negotiation, but from my experience we need negotiation.

**Comment by matejvasek** [comment]
@adambkaplan wort case: I can use the old deprecated ctor and override linter error.

**Comment by matejvasek** [comment]
PTAL @adambkaplan

File: pkg/docker/docker.go
Comment (matejvasek) [reply pkg/docker/docker.go:325]: IMO this is expected and desired behaviour.

File: pkg/docker/docker.go
Comment (adambkaplan) [inline pkg/docker/docker.go:322]: It took me a long time to fully understand how the CLI options interact - I'm recommend we add a comment for future maintainers.

```suggestion
        // Create a new docker client with the provided host endpoint and HTTP Client.
        // By default, this client will negotiate the Docker API version to use with the backend engine on the first request.
        // The API version to use can be fixed by setting the DOCKER_API_VERSION environment variable.
        // See https://pkg.go.dev/github.com/docker/docker/client#WithAPIVersionNegotiation and
        // https://pkg.go.dev/github.com/docker/docker/client#WithVersionFromEnv for more information
	return dockerapi.NewClientWithOpts(
```

**Review by adambkaplan — CHANGES_REQUESTED** [review]
One last suggestion, otherwise looks good.

What started out as a chore feels like a net new feature! I don't think it warrants any additional documentation beyond a release note, but after this lands we should increase the minor version number on the next release (to 1.4.0).

**Comment by adambkaplan** [comment]
/assign @sayan-biswas 

Feel free to use `/approve` to green-light merge - I can add the LGTM once the comment suggestion is added commits are squashed.

@matejvasek - in your commit message please explain why we need API version negotiation for docker. This makes it easier for future maintainers to understand why the change was introduced, without having to chase down this pull request and unroll the comments.

**Pull request merged**
